Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm

```diff
@@ -0,0 +1,114 @@
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+
+#import "config.h"
+#import "WebExtensionDynamicScripts.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+#import "WKContentWorld.h"
+#import "WKFrameInfoPrivate.h"
+#import "WKWebViewPrivate.h"
+#import "WebExtensionContext.h"
+#import "WebExtensionUtilities.h"
+#import "_WKFrameHandle.h"
+#import "_WKFrameTreeNode.h"
+#import <wtf/URL.h>
+
+namespace WebKit {
+
+namespace DynamicScripts {
+
+void addResultAndCompleteIfDone(id resultOfJSExecution, WKFrameInfo *frame, NSString *errorMessage, size_t numberOfExecutions, Vector<WebExtensionScriptInjectionResultParameters>&& results, CompletionHandler<void(std::optional<Vector<WebExtensionScriptInjectionResultParameters>>)>&& callback)
+{
+    WebExtensionScriptInjectionResultParameters result;
+    result.result = resultOfJSExecution ?: @"";
+    result.frameID = frame.isMainFrame ? mainFrameID : frame._handle.frameID;
+    if (errorMessage.length)
+        result.error = errorMessage;
+
+    results.append(result);
+    if (results.size() == numberOfExecutions)
+        callback(results);
+}
+
+void populateFrames(_WKFrameTreeNode *currentNode, size_t startingFrameID, bool allFrames, bool foundStartingFrame, Vector<RetainPtr<_WKFrameTreeNode>>&& frames)
+{
+    if (foundStartingFrame)
+        frames.append(WTFMove(currentNode));
+
+    WKFrameInfo *info = currentNode.info;
+    size_t currentFrameID = info.isMainFrame ? mainFrameID : info._handle.frameID;
+    if (currentFrameID == startingFrameID) {
+        frames.append(WTFMove(currentNode));
+        if (!allFrames)
+            return;
+
+        foundStartingFrame = true;
+    }
+
+    for (_WKFrameTreeNode *child in currentNode.childFrames)
+        DynamicScripts::populateFrames(child, startingFrameID, allFrames, foundStartingFrame, WTFMove(frames));
+}
+
+void populateFrames(_WKFrameTreeNode *currentNode, std::optional<Vector<size_t>> frameIDs, Vector<RetainPtr<_WKFrameTreeNode>>&& frames)
+{
+    WKFrameInfo *info = currentNode.info;
+    size_t currentFrameID = info.isMainFrame ? mainFrameID : info._handle.frameID;
+    if (frameIDs && frameIDs->contains(currentFrameID))
+        frames.append(WTFMove(currentNode));
+
+    for (_WKFrameTreeNode *child in currentNode.childFrames)
+        populateFrames(child, WTFMove(frameIDs), WTFMove(frames));
+}
+
+std::optional<SourcePair> sourcePairForResource(String path, RefPtr<WebExtension> extension)
+{
+    NSData *scriptData = extension->resourceDataForPath(path);
+    if (!scriptData)
+        return std::nullopt;
+
+    URL resourceURL = URL(extension->resourceFileURLForPath(path));
+    return SourcePair([[NSString alloc] initWithData:scriptData encoding:NSUTF8StringEncoding], resourceURL);
+}
+
+void populateStyleSheetPairs(StyleSheetPairs&& pairs, RefPtr<WebExtension> extension, std::optional<Vector<String>> files, std::optional<String> css)
```

Comment:
Ah I see whats going on. You want to take `pairs` by mutable reference, not by rvalue ref.
Even better: return the StyleSheetPairs from this function instead of taking as a parameter and mutating.

Reply:
Also ditto the std::optional<String> comment from before.